woocommerce/woocommerce-android
# Add support for fetching custom order stats

Fixes #528 

This PR adds adds a new option to the order stats card on the **Store** tab for user to add a tab and select a custom date range and granularity. 

### User Flow
* User launches the app and clicks on CUSTOM tab
* If no previous date selection is available, a dialog is displayed and user selects a custom date range and granularity (how the data should be displayed). Once selected data is fetched and displayed.
* If previous date selection is available, data is fetched for this date range and displayed.
* User has the option to update the date range and granularity at any time.

### How it works
When custom tab is selected, the fluxC db is queried for the `WCOrderStatsModel` with the `isCustomField` flag set to true. If none exists, a new `FetchOrderStatsPayload` is created with the custom date range and granularity and a request is fired to fetch new stats from the api. When the `OnWCStatsChanged` event is fired, the custom tab is updated with the values from the db. 

The stats will be refreshed:
* on app startup, when custom tab is selected
* during “Pull to refresh” list gesture, when custom tab is selected

### Screenshots
<img src="https://user-images.githubusercontent.com/22608780/52857510-6946c800-314d-11e9-8cd8-7b0c3542e5e5.png" width="300"/>

<img src="https://user-images.githubusercontent.com/22608780/52834496-2b738080-3108-11e9-9d0b-fcee75141b67.png" width="300"/>

### Notes
The [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1103) must be merged before this can be merged

### Tests
Unit tests and instrumentation tests are included in `DashboardPresenterTest` . Manually tested the following scenarios:

1. Select custom TAB:
    * Expected behavior
       * **when app is first launched:** Dialog opened with **From** and **To** set to current date. The **Display In** set to DAYS. When **OK** is pressed, skeleton view is displayed until data is fetched from FluxC.
           * Success Scenario: Data is displayed. 
           * If no internet available: Error view displayed with **No data available** message.
           * Error Scenario: Error image is displayed.
       * **when app is relaunched second time**: Data is fetched from local cache for previously selected data range and granularity.
       * **when pull-to-refresh gesture**: Data is fetched from api for previously selected data range and granularity. When **OK** is pressed, skeleton view is displayed until data is fetched from FluxC.
           * Success Scenario: Data is displayed. 
           * If no internet available: Error view displayed with **No data available** message.
           * Error Scenario: Error image is displayed.
2. Click on Edit icon in CUSTOM Tab:
    * Expected behavior
       * Dialog opened with **From** and **To** and **Display In** set to previously selected values. When **OK** is pressed, skeleton view is displayed until data is fetched from FluxC.
           * Success Scenario: Data is displayed. 
           * If no internet available: Error view displayed with **No data available** message.
           * Error Scenario: Error image is displayed.


Update release notes:
